theano/gpuarray/opt.py

```diff
@@ -319,7 +319,8 @@ def apply(self, fgraph):
 
             context_name = None
             for i in [mapping[i] for i in node.inputs]:
-                if isinstance(i.type, GpuArrayType):
+                target = getattr(i.tag, 'target', None)
+                if (isinstance(i.type, GpuArrayType) and target != 'cpu'):
```

Comment:
This is not the right check since target could be `'gpu'` and that means the old backend.

In any case the current code does not support target in the middle of the graph.  The right way to do it is to call .transfer() which will explicitly insert a GpuToGpu node.  It is supported for the inputs though.